Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift

```diff
@@ -0,0 +1,297 @@
+///          },
+///          {
+///             "name":"version",
+///             "type":"string"
+///          },
+///          {
+///             "name":"chainId",
+///             "type":"uint256"
+///          },
+///          {
+///             "name":"verifyingContract",
+///             "type":"address"
+///          }
+///       ],
+///       "Person":[
+///          {
+///             "name":"name",
+///             "type":"string"
+///          },
+///          {
+///             "name":"wallet",
+///             "type":"address"
+///          }
+///       ],
+///       "Mail":[
+///          {
+///             "name":"from",
+///             "type":"Person"
+///          },
+///          {
+///             "name":"to",
+///             "type":"Person"
+///          },
+///          {
+///             "name":"contents",
+///             "type":"string"
+///          }
+///       ]
+///    },
+///    "primaryType":"Mail",
+///    "domain":{
+///       "name":"Ether Mail",
+///       "version":"1",
+///       "chainId":1,
+///       "verifyingContract":"0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC"
+///    },
+///    "message":{
+///       "from":{
+///          "name":"Cow",
+///          "wallet":"0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"
+///       },
+///       "to":{
+///          "name":"Bob",
+///          "wallet":"0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"
+///       },
+///       "contents":"Hello, Bob!"
+///    }
+/// }
+/// ```
+///
+/// Example use case:
+/// ```
+///     let payload: String = ... // This is the payload received from eth_signTypedData
+///     let eip712TypedData = try EIP712Parser.parse(payload)
+///     let signature = try Web3Signer.signEIP712(
+///         eip712TypedData,
+///         keystore: keystore,
+///         account: account,
+///         password: password)
+/// ```
+public class EIP712Parser {
+
+    static func toData(_ json: String) throws -> Data {
+        guard let json = json.data(using: .utf8) else {
+            throw Web3Error.inputError(desc: "EIP712Parser. Failed to parse EIP712 payload. Given string is not valid UTF8 string.")
+        }
+        return json
+    }
+
+    public static func parse(_ rawJson: String) throws -> EIP712TypedData {
+        try parse(try toData(rawJson))
+    }
+
+    public static func parse(_ rawJson: Data) throws -> EIP712TypedData {
+        let decoder = JSONDecoder()
+        let types = try decoder.decode(EIP712TypeArray.self, from: rawJson).types
+        guard let json = try rawJson.asJsonDictionary() else {
+            throw Web3Error.inputError(desc: "EIP712Parser. Cannot decode given JSON as it cannot be represented as a Dictionary. Is it valid JSON?")
+        }
+        guard let primaryType = json["primaryType"] as? String else {
+            throw Web3Error.inputError(desc: "EIP712Parser. Top-level string field 'primaryType' missing.")
+        }
+        guard let domain = json["domain"] as? [String : AnyObject] else {
```

Comment:
Maybe we should add here some at least short reasoning about why it's left here as a dictionary, rather being casted in a concrete type?